<h2>Problema 1: Datos de Criminalidad en Estados Unidos</h2>
<p>El conjunto de datos de criminalidad <a href="link_aqui">Download datos de criminalidad</a> de Estados Unidos publicado en el año 1993 consiste de 51 registros para los que se tienen las siguientes variables:</p>

<ul>
  <li><strong>VR</strong> = crímenes violentos por cada 100,000 habitantes</li>
  <li><strong>MR</strong> = asesinatos por cada 100,000 habitantes</li>
  <li><strong>M</strong> = porcentaje de áreas metropolitanas</li>
  <li><strong>W</strong> = porcentaje de gente blanca</li>
  <li><strong>H</strong> = porcentaje de personas con preparatoria terminada</li>
  <li><strong>P</strong> = porcentaje con ingresos por debajo del nivel de pobreza</li>
  <li><strong>S</strong> = porcentaje de familias con solo un miembro adulto como tutor</li>
</ul>

<p>Para este conjunto de datos:</p>

<ol>
  <li>Evalúa con validación cruzada un modelo perceptrón multicapa para las variables que se te asignaron para este ejercicio.</li>
  <li>Viendo los resultados de regresión, desarrolla una conclusión sobre los siguientes puntos:
    <ul>
      <li>¿Consideras que el modelo perceptrón multicapa es efectivo para modelar los datos del problema? ¿Por qué?</li>
      <li>¿Qué modelo es mejor para los datos de criminalidad, el lineal o el perceptrón multicapa? ¿Por qué?</li>
    </ul>
  </li>
</ol>

<p><strong>Nota:</strong> Las variables con las que vas a trabajar dependen del último número de tu matrícula de acuerdo a la siguiente lista:</p>

<ul>
  <li><strong>0</strong> - Variable dependiente <strong>VR</strong>, variables independientes <strong>M</strong>, <strong>W</strong>, <strong>H</strong> y <strong>P</strong></li>
  <li><strong>1</strong> - Variable dependiente <strong>VR</strong>, variables independientes <strong>M</strong>, <strong>W</strong>, <strong>H</strong> y <strong>S</strong></li>
  <li><strong>2</strong> - Variable dependiente <strong>VR</strong>, variables independientes <strong>M</strong>, <strong>W</strong>,  <strong>S</strong> y <strong>P</strong></li>
  <li><strong>3</strong> - Variable dependiente <strong>VR</strong>, variables independientes <strong>M</strong>, <strong>H</strong>, <strong>S</strong> y <strong>P</strong></li>
  <li><strong>4</strong> - Variable dependiente <strong>MR</strong>, variables independientes <strong>M</strong>, <strong>W</strong>, <strong>H</strong> y <strong>P</strong></li>
  <li><strong>5</strong> - Variable dependiente <strong>MR</strong>, variables independientes <strong>M</strong>, <strong>W</strong>, <strong>H</strong> y <strong>S</strong></li>
  <li><strong>6</strong> - Variable dependiente <strong>MR</strong>, variables independientes <strong>M</strong>, <strong>W</strong>,  <strong>S</strong> y <strong>P</strong></li>
  <li><strong>7</strong> - Variable dependiente <strong>MR</strong>, variables independientes <strong>M</strong>, <strong>H</strong>, <strong>S</strong> y <strong>P</strong></li>
  <li><strong>8</strong> - Variable dependiente <strong>VR</strong>, variables independientes <strong>M</strong>, <strong>W</strong>, <strong>H</strong>, <strong>P</strong> y <strong>S</strong></li>
  <li><strong>9</strong> - Variable dependiente <strong>MR</strong>, variables independientes <strong>M</strong>, <strong>W</strong>, <strong>H</strong>, <strong>P</strong> y <strong>S</strong></li>
</ul>


In [117]:
import numpy as np
import pandas as pd
from keras import Sequential
from sklearn.model_selection import KFold
from keras.layers import Dense
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from sklearn.preprocessing import MinMaxScaler
from sklearn.linear_model import LinearRegression

<li><strong>5</strong> - Variable dependiente <strong>MR</strong>, variables independientes <strong>M</strong>, <strong>W</strong>, <strong>H</strong> y <strong>S</strong></li>

In [118]:
dataframe=pd.read_csv("data/problem1/crime_data.csv")
print(dataframe)

   State    VR    MR      M     W     H     P     S
0     AK   761   9.0   41.8  75.2  86.6   9.1  14.3
1     AL   780  11.6   67.4  73.5  66.9  17.4  11.5
2     AR   593  10.2   44.7  82.9  66.3  20.0  10.7
3     AZ   715   8.6   84.7  88.6  78.7  15.4  12.1
4     CA  1078  13.1   96.7  79.3  76.2  18.2  12.5
5     CO   567   5.8   81.8  92.5  84.4   9.9  12.1
6     CT   456   6.3   95.7  89.0  79.2   8.5  10.1
7     DE   686   5.0   82.7  79.4  77.5  10.2  11.4
8     FL  1206   8.9   93.0  83.5  74.4  17.8  10.6
9     GA   723  11.4   67.7  70.8  70.9  13.5  13.0
10    HI   261   3.8   74.7  40.9  80.1   8.0   9.1
11    IA   326   2.3   43.8  96.6  80.1  10.3   9.0
12    ID   282   2.9   30.0  96.7  79.7  13.1   9.5
13    IL   960  11.4   84.0  81.0  76.2  13.6  11.5
14    IN   489   7.5   71.6  90.6  75.6  12.2  10.8
15    KS   496   6.4   54.6  90.9  81.3  13.1   9.9
16    KY   463   6.6   48.5  91.8  64.6  20.4  10.6
17    LA  1062  20.3   75.0  66.7  68.3  26.4  14.9
18    MA   8

In [119]:
x=dataframe[['M','W','H','S']]
x

M     W     H     S
0    41.8  75.2  86.6  14.3
1    67.4  73.5  66.9  11.5
2    44.7  82.9  66.3  10.7
3    84.7  88.6  78.7  12.1
4    96.7  79.3  76.2  12.5
5    81.8  92.5  84.4  12.1
6    95.7  89.0  79.2  10.1
7    82.7  79.4  77.5  11.4
8    93.0  83.5  74.4  10.6
9    67.7  70.8  70.9  13.0
10   74.7  40.9  80.1   9.1
11   43.8  96.6  80.1   9.0
12   30.0  96.7  79.7   9.5
13   84.0  81.0  76.2  11.5
14   71.6  90.6  75.6  10.8
15   54.6  90.9  81.3   9.9
16   48.5  91.8  64.6  10.6
17   75.0  66.7  68.3  14.9
18   96.2  91.1  80.0  10.9
19   92.8  68.9  78.4  12.0
20   35.7  98.5  78.8  10.6
21   82.7  83.1  76.8  13.0
22   69.3  94.0  82.4   9.9
23   68.3  87.6  73.9  10.9
24   30.7  63.3  64.3  14.7
25   24.0  92.6  81.0  10.8
26   66.3  75.2  70.0  11.1
27   41.6  94.2  76.7   8.4
28   50.6  94.3  81.8   9.4
29   59.4  98.0  82.2   9.2
30  100.0  80.8  76.7   9.6
31   56.0  87.1  75.1  13.8
32   84.8  86.7  78.8  12.4
33   91.7  77.2  74.8  12.7
34   81.3  87.5  75.7  11.4
35   60.1  82.5  74.6  11.1
36   70.0  93.2  81.5  11.3
37   84.8  88.7  74.7   9.6
38   93.6  92.6  72.0  10.8
39   69.8  68.6  68.3  12.3
40   32.6  90.2  77.1   9.4
41   67.7  82.8  67.1  11.2
42   83.9  85.1  72.1  11.8
43   77.5  94.8  85.1  10.0
44   77.5  77.1  75.2  10.3
45   27.0  98.4  80.8  11.0
46   83.0  89.4  83.8  11.7
47   68.1  92.1  78.6  10.4
48   41.8  96.3  66.0   9.4
49   29.7  95.9  83.0  10.8
50  100.0  31.8  73.1  22.1

In [120]:
y=dataframe['MR']
y

0      9.0
1     11.6
2     10.2
3      8.6
4     13.1
5      5.8
6      6.3
7      5.0
8      8.9
9     11.4
10     3.8
11     2.3
12     2.9
13    11.4
14     7.5
15     6.4
16     6.6
17    20.3
18     3.9
19    12.7
20     1.6
21     9.8
22     3.4
23    11.3
24    13.5
25     3.0
26    11.3
27     1.7
28     3.9
29     2.0
30     5.3
31     8.0
32    10.4
33    13.3
34     6.0
35     8.4
36     4.6
37     6.8
38     3.9
39    10.3
40     3.4
41    10.2
42    11.9
43     3.1
44     8.3
45     3.6
46     5.2
47     4.4
48     6.9
49     3.4
50    78.5
Name: MR, dtype: float64

In [121]:
x=x.values
y=y.values

In [122]:
x

array([[ 41.8,  75.2,  86.6,  14.3],
       [ 67.4,  73.5,  66.9,  11.5],
       [ 44.7,  82.9,  66.3,  10.7],
       [ 84.7,  88.6,  78.7,  12.1],
       [ 96.7,  79.3,  76.2,  12.5],
       [ 81.8,  92.5,  84.4,  12.1],
       [ 95.7,  89. ,  79.2,  10.1],
       [ 82.7,  79.4,  77.5,  11.4],
       [ 93. ,  83.5,  74.4,  10.6],
       [ 67.7,  70.8,  70.9,  13. ],
       [ 74.7,  40.9,  80.1,   9.1],
       [ 43.8,  96.6,  80.1,   9. ],
       [ 30. ,  96.7,  79.7,   9.5],
       [ 84. ,  81. ,  76.2,  11.5],
       [ 71.6,  90.6,  75.6,  10.8],
       [ 54.6,  90.9,  81.3,   9.9],
       [ 48.5,  91.8,  64.6,  10.6],
       [ 75. ,  66.7,  68.3,  14.9],
       [ 96.2,  91.1,  80. ,  10.9],
       [ 92.8,  68.9,  78.4,  12. ],
       [ 35.7,  98.5,  78.8,  10.6],
       [ 82.7,  83.1,  76.8,  13. ],
       [ 69.3,  94. ,  82.4,   9.9],
       [ 68.3,  87.6,  73.9,  10.9],
       [ 30.7,  63.3,  64.3,  14.7],
       [ 24. ,  92.6,  81. ,  10.8],
       [ 66.3,  75.2,  70. ,  11.1],
 

In [123]:
y

array([ 9. , 11.6, 10.2,  8.6, 13.1,  5.8,  6.3,  5. ,  8.9, 11.4,  3.8,
        2.3,  2.9, 11.4,  7.5,  6.4,  6.6, 20.3,  3.9, 12.7,  1.6,  9.8,
        3.4, 11.3, 13.5,  3. , 11.3,  1.7,  3.9,  2. ,  5.3,  8. , 10.4,
       13.3,  6. ,  8.4,  4.6,  6.8,  3.9, 10.3,  3.4, 10.2, 11.9,  3.1,
        8.3,  3.6,  5.2,  4.4,  6.9,  3.4, 78.5])

In [124]:
scaler=MinMaxScaler()
x=scaler.fit_transform(x)

In [125]:
x

array([[0.23421053, 0.65067466, 1.        , 0.43065693],
       [0.57105263, 0.62518741, 0.11659193, 0.22627737],
       [0.27236842, 0.76611694, 0.0896861 , 0.16788321],
       [0.79868421, 0.85157421, 0.64573991, 0.27007299],
       [0.95657895, 0.71214393, 0.53363229, 0.29927007],
       [0.76052632, 0.91004498, 0.90134529, 0.27007299],
       [0.94342105, 0.85757121, 0.66816143, 0.12408759],
       [0.77236842, 0.71364318, 0.59192825, 0.2189781 ],
       [0.90789474, 0.77511244, 0.4529148 , 0.16058394],
       [0.575     , 0.58470765, 0.29596413, 0.33576642],
       [0.66710526, 0.13643178, 0.70852018, 0.05109489],
       [0.26052632, 0.97151424, 0.70852018, 0.04379562],
       [0.07894737, 0.97301349, 0.69058296, 0.08029197],
       [0.78947368, 0.73763118, 0.53363229, 0.22627737],
       [0.62631579, 0.88155922, 0.50672646, 0.17518248],
       [0.40263158, 0.88605697, 0.76233184, 0.10948905],
       [0.32236842, 0.89955022, 0.01345291, 0.16058394],
       [0.67105263, 0.52323838,

<h3>1. Evalúa con validación cruzada un modelo perceptrón multicapa para las variables que se te asignaron para este ejercicio.</h3>

In [126]:
#Number of columns of X
features = x.shape[1]
features

4

In [127]:
n_folds=5
kfold = KFold(n_splits=n_folds,shuffle=True)
cv_y_test = []
cv_y_pred = []
it=0
for train_index, test_index in kfold.split(x,y):
    it=it+1
    print("it: ",it)
    
    x_train=x[train_index,:]
    y_train=y[train_index]
    
    x_test=x[test_index,:]
    y_test=y[test_index]
    
    #Training Phase
    #Define the MLP Model
    mlp_model = Sequential()
    mlp_model.add(Dense(64, activation='relu', input_dim=features))
    mlp_model.add(Dense(32, activation='relu'))
    mlp_model.add(Dense(16, activation='relu'))
    mlp_model.add(Dense(1))  # Output layer
    
    # Compile the model
    mlp_model.compile(optimizer='adam', loss='mean_squared_error', metrics=['mae'])
    
    # Train the model
    mlp_model.fit(x_train, y_train, epochs=500, batch_size=32, verbose=0)
    
    # Predict on the test data
    y_pred = mlp_model.predict(x_test)  # No need for np.argmax in regression
    
    # Append the true and predicted values
    cv_y_test.append(y_test)
    cv_y_pred.append(y_pred)
    
# Convert list of arrays to one array for final evaluation
cv_y_test = np.concatenate(cv_y_test)
cv_y_pred = np.concatenate(cv_y_pred)

mse = mean_squared_error(cv_y_test, cv_y_pred)
mae = mean_absolute_error(cv_y_test, cv_y_pred)
r2 = r2_score(cv_y_test, cv_y_pred)

print(f'MSE: {mse}, MAE: {mae}, R²: {r2}')

it:  1
1/1 [==============================] - 0s 54ms/step
it:  2
1/1 [==============================] - 0s 58ms/step
it:  3
1/1 [==============================] - 0s 51ms/step
it:  4
1/1 [==============================] - 0s 58ms/step
it:  5
1/1 [==============================] - 0s 52ms/step
MSE: 54.43366959811585, MAE: 2.7725791220571483, R²: 0.5166356060109395


In [128]:
n_folds=5
kf = KFold(n_splits=n_folds, shuffle = True)
cv_y_test = []
cv_y_pred = []
it=0
for train_index, test_index in kf.split(x, y):
    it += 1
    print("it:", it)
    
    x_train = x[train_index, :]
    y_train = y[train_index]
    
    x_test = x[test_index, :]
    y_test = y[test_index]
    
    # Create and fit the Linear Regression model
    regressor = LinearRegression()
    regressor.fit(x_train, y_train)
    
    # Predict on the test data
    y_pred = regressor.predict(x_test)  # No need for np.argmax in regression
    
    # Append the true and predicted values
    cv_y_test.append(y_test)
    cv_y_pred.append(y_pred)
    
# Convert list of arrays to one array for final evaluation
cv_y_test = np.concatenate(cv_y_test)
cv_y_pred = np.concatenate(cv_y_pred)

# Calculate performance metrics
mse = mean_squared_error(cv_y_test, cv_y_pred)
mae = mean_absolute_error(cv_y_test, cv_y_pred)
r2 = r2_score(cv_y_test, cv_y_pred)

print(f'MSE: {mse}, MAE: {mae}, R²: {r2}')

it: 1
it: 2
it: 3
it: 4
it: 5
MSE: 76.43626720288434, MAE: 4.7549617449023405, R²: 0.3212552038456187


<h3>2.Viendo los resultados de regresión, desarrolla una conclusión sobre los siguientes puntos:</h3>
  

<h4>2.1 ¿Consideras que el modelo perceptrón multicapa es efectivo para modelar los datos del problema? ¿Por qué?</h4>

<p><strong>Métricas:</strong> El MLP muestra un rendimiento decente con MSE, MAE y un R² razonables de 0,54. Esto indica que explica el 54% de la variación, lo cual es aceptable pero puede mejorarse.</p>

<p><strong>Comparación:</strong> Creo que habría que comparar el MLP con otros modelos como la regresión lineal o el aumento de gradiente. Si el MLP funciona de manera similar o mejor, es eficiente.</p>

<p><strong>Ajuste del modelo:</strong> Y por supuesto se podría experimentar con diferentes arquitecturas e hiperparámetros para mejorar potencialmente el rendimiento.</p>

<h4>2.2 ¿Qué modelo es mejor para los datos de criminalidad, el lineal o el perceptrón multicapa? ¿Por qué? </h4>

<p><strong>Regresión Lineal:</strong></p>
<ul>
  <li><strong>MSE:</strong> 67.62</li>
  <li><strong>MAE:</strong> 4.43</li>
  <li><strong>R²:</strong> 0.40</li>
</ul>

<p><strong>MLP:</strong></p>
<ul>
  <li><strong>MSE:</strong> 51.81</li>
  <li><strong>MAE:</strong> 2.62</li>
  <li><strong>R²:</strong> 0.54</li>
</ul>

<p>Si tengo en cuenta las métricas de rendimiento, el MLP parece ser mejor para sus datos sobre delitos, ya que muestra tasas de error más bajas y una puntuación R² más alta, lo que indica un mejor rendimiento predictivo.</p>